Modules documentation (#1)
andrewbranch/docs: modules/01_Theory.md

# Theory

## Scripts and modules in JavaScript

In the early days of JavaScript, when the language only ran in browsers, there were no modules, but it was still possible to split the JavaScript for a web page into multiple files by using multiple `script` tags in HTML:

```html
<html>
  <head>
    <script src="a.js"></script>
    <script src="b.js"></script>
  </head>
  <body></body>
</html>
```

> This approach had some downsides, especially as web pages grew larger and more complex. In particular, all scripts loaded onto the same page share the same scope—appropriately called the “global scope”—meaning the scripts had to be very careful not to overwrite each others’ variables and functions.

Reviewer: We might need to define "scope" more explicitly; not sure

>
> ```js
> // a.js
> exports.message = "Hello from a.js";
> ```
>
> ```js
> // b.js
> const a = require("./a");
> console.log(a.message); // 'Hello from a.js'
> ```

Accordingly, when TypeScript detects that a file is a CommonJS or ECMAScript module, it starts by assuming that file will have its own scope. Beyond that, though, the compiler’s job gets a little more complicated.

## TypeScript’s job concerning modules

The TypeScript compiler’s chief goal is to look at input code and tell the author about problems the output code might encounter at runtime. To do this, the compiler needs to know some things about the code’s intended runtime environment—what globals are available, for example. When the input code is a module, there are several additional questions the compiler needs to answer in order to do its job. Let’s use a few lines of input code as an example to think about all the information needed to analyze it:

```ts
import sayHello from "greetings";
sayHello("world");
```

> 

Reviewer: This section is good but sort of puts the cart before the horse. We can start with "At runtime, something will happen" (via all these same steps described in checking), then describe TS's job in terms of asking questions about what that behavior is.

Author: Hm, that’s a good point, but this is supposed to be `.ts` file contents, so it’s hard to say what’s going to happen at runtime without presupposing what module format we’re going to emit. But maybe it can be stated abstractly enough to gloss over the potential ESM/CJS differences here.

Author: The [updated version](https://delightful-glacier-09aa73710-2946.centralus.azurestaticapps.net/docs/handbook/modules/theory.html#typescripts-job-concerning-modules) of this section was rewritten to account for the possibility of a runtime that may load the TS file directly, which further complicates prefacing this with a discussion of what the runtime will do 🤔

- [**`es2015`**](./03_Reference.md#es2015-es2020-es2022-esnext): Reflects the ES2015 language specification for JavaScript modules (the version that first introduced `import` and `export` to the language).
- [**`es2020`**](./03_Reference.md#es2015-es2020-es2022-esnext): Adds support for `import.meta` and `export * as ns from "mod"` to `es2015`.
- [**`es2022`**](./03_Reference.md#es2015-es2020-es2022-esnext): Adds support for top-level `await` to `es2020`.
- [**`esnext`**](./03_Reference.md#es2015-es2020-es2022-esnext): Currently identical to `es2022`, but will be a moving target reflecting the latest ECMAScript specifications, as well as module-related Stage 3+ proposals that are expected to be included in upcoming specification versions.
- **[`commonjs`](./03_Reference.md#commonjs), [`system`](./03_Reference.md#system), [`amd`](./03_Reference.md#amd), and [`umd`](./03_Reference.md#umd)**: Each emits everything in the non-standard module system named, and assumes everything can be successfully imported into that module system. These are no longer recommended for new projects and will not be covered by this documentation.

> 

Reviewer: `"none"`

>
> The upside of this “author ESM, output anything” strategy was that TypeScript could use standard JavaScript syntax, making the authoring experience familiar to newcomers, and (theoretically) making it easy for projects to start targeting ESM outputs in the future. There are three significant downsides, which became fully apparent only after ESM and CJS modules were allowed to coexist and interoperate in Node.js:
>

> > 1. Early assumptions about how ESM/CJS interoperability would work in Node.js were wrong, and today, interoperability rules differ between Node.js and bundlers. Consequently, the configuration space for modules in TypeScript is large.

Reviewer: ```suggestion
> 1. Early assumptions about how ESM/CJS interoperability would work in Node.js turned out to be wrong, and today, interoperability rules differ between Node.js and bundlers. Consequently, the configuration space for modules in TypeScript is large.
```

>
> In TypeScript 5.0, a new compiler option called `verbatimModuleSyntax` was introduced to help TypeScript authors know exactly how their `import` and `export` statements will be emitted. When enabled, the flag requires imports and exports in input files to be written in the form that will undergo the least amount of transformation before emit. So if a file will be emitted as ESM, imports and exports must be written in ESM syntax; if a file will be emitted as CJS, it must be written in the CommonJS-inspired TypeScript syntax (`import fs = require("fs")` and `export = {}`). This setting is particularly recommended for Node.js projects that use mostly ESM, but have a select few CJS files. It is not recommended for projects that currently target CJS, but may want to target ESM in the future.

### ESM and CJS interoperability

> Can an ES module `import` a CommonJS module? If so, does a default import select `exports` or `exports.default`? Can a CommonJS module `require` an ES module? CommonJS isn’t part of the ECMAScript specification, so runtimes, bundlers, and transpilers have been free to make up their own answers to these questions since ESM was standardized in 2015. Today, interoperability rules between ESM and CJS for most runtimes and bundlers broadly fall into one of three categories:

Reviewer: ```suggestion
Can an ES module `import` a CommonJS module? If so, does a default import select `exports` or `exports.default`? Can a CommonJS module `require` an ES module? CommonJS isn’t part of the ECMAScript specification, so runtimes, bundlers, and transpilers have been free to make up their own answers to these questions since ESM was standardized in 2015, and as such no standard set of interop rules exist. Today, interoperability rules between ESM and CJS for most runtimes and bundlers broadly fall into one of three categories:
```

While the `module` compiler option can transform imports and exports in input files to different module formats in output files, the module _specifier_ (the string `from` which you `import`, or pass to `require`) is always emitted as-written. For example, an input like:

```ts
import { add } from "./math.mjs";
add(1, 2);
```

might be emitted as either:

```ts
import { add } from "./math.mjs";
add(1, 2);
```

or:

```ts
const math_1 = require("./math.mjs");
math_1.add(1, 2);
```

depending on the `module` compiler option, but the module specifier will always be `"./math.mjs"`. There is no compiler option that enables transforming, substituting, or rewriting module specifiers. Consequently, module specifiers must be written in a way that works for the code’s target runtime or bundler, and it’s TypeScript’s job to understand those _output_-relative specifiers. The process of finding the file referenced by a module specifier is called _module resolution_.

## Module resolution

Let’s return to our [first example](#typescripts-job-concerning-modules) and review what we’ve learned about it so far:

```ts
import sayHello from "greetings";
sayHello("world");
```

> So far, we’ve discussed how the host’s module system and TypeScript’s `module` compiler option might impact this code: we know that the input syntax looks like ESM, but the output format depends on the `module` compiler option and potentially the file extension and `package.json` `"type"` field; we also know that what `sayHello` gets bound to, and even whether the import is even allowed, may vary depending on the module kinds of this file and the target file. But we haven’t yet discussed how to _find_ the target file.

Reviewer: ```suggestion
So far, we’ve discussed how the host’s module system and TypeScript’s `module` compiler option might impact this code. We know that the input syntax looks like ESM, but the output format depends on the `module` compiler option, potentially the file extension, and `package.json` `"type"` field. We also know that what `sayHello` gets bound to, and even whether the import is even allowed, may vary depending on the module kinds of this file and the target file. But we haven’t yet discussed how to _find_ the target file.
```

### Module resolution is host-defined

While the ECMAScript specification defines how to parse and interpret `import` and `export` statements, it leaves module resolution up to the host. If you’re creating a hot new JavaScript runtime, you’re free to create a module resolution scheme like:

```ts
import monkey from "🐒"; // Looks for './eats/bananas.js'
import cow from "🐄";    // Looks for './eats/grass.js'
import lion from "🦁";   // Looks for './eats/you.js'
```

> and still claim to implement “standards-compliant ESM.” Needless to say, TypeScript would have no idea what types to assign to `monkey`, `cow`, and `lion` without built-in knowledge of this runtime’s module resolution algorithm. Just as `module` informs the compiler about the host’s expected module format, `moduleResolution`, along with a few customization options, specify the algorithm the host uses to resolve module specifiers to files.

Reviewer: ```suggestion
and still claim to implement “standards-compliant ESM.” Needless to say, TypeScript would have no idea what types to assign to `monkey`, `cow`, and `lion` without built-in knowledge of this runtime’s module resolution algorithm. Just as `module` informs the compiler about the host’s expected module format, `moduleResolution`, along with a few customization options, specify the algorithm the host uses to resolve module specifiers to files. This also clarifies why TypeScript doesn't modify import specifiers during emit: the relationship between an import specifier and a file on disk (if one even exists) is host-defined, and TypeScript is not a host.
```

The available `moduleResolution` options are:

> - [**`classic`**](./03_Reference.md#classic): TypeScript’s oldest and most inscrutable module resolution mode, this is unfortunately the default when `module` is set to anything other than `commonjs`, `node16`, or `nodenext`. It does not represent any real-world module resolver and should never be used. It was kept around only for backwards compatibility.

Reviewer: ```suggestion
- [**`classic`**](./03_Reference.md#classic): TypeScript’s oldest and most inscrutable module resolution mode, this is unfortunately the default when `module` is set to anything other than `commonjs`, `node16`, or `nodenext`. It does not represent any real-world module resolver and should never be used. It was kept around only for backwards compatibility. This option is scheduled for deprecation in TypeScript 6.0.
```

Author: I took a marginally higher effort stab at figuring out why this exists in some new edits on the website PR, but will add the deprecation note as well:

> TypeScript’s oldest module resolution mode, this is unfortunately the default when `module` is set to anything other than `commonjs`, `node16`, or `nodenext`. It was probably made to provide best-effort resolution for a wide range of [RequireJS](https://requirejs.org/docs/api.html#packages) configurations. It should not be used for new projects (or even old projects that don’t use RequireJS or another AMD module loader).

3. to know what **type** to assign to **imported names**.

## Who is the host?

Before we move on, it’s worth making sure we’re on the same page about the term _host_, because it will come up frequently. We defined it before as “the system that ultimately consumes the output code to direct its module loading behavior.” In other words, it’s the system outside of TypeScript that TypeScript’s module analysis tries to model:

- When the output code (whether produced by `tsc` or a third-party transpiler) is run directly in a runtime like Node.js, the runtime is the host.
- When there is no “output code” because a runtime consumes TypeScript files directly, the runtime is still the host.
- When a bundler consumes TypeScript inputs or outputs and produces a bundle, the bundler is the host, because it looked at the original set of imports/requires, looked up what files they referenced, and produced a new file or set of files where the original imports and requires are erased or transformed beyond recognition. (That bundle itself might comprise modules, and the runtime that runs it will be its host, but TypeScript doesn’t know about anything that happens post-bundler.)

> - If another transpiler, optimizer, or formatter runs on TypeScript’s outputs, it’s _not_ a host that TypeScript cares about, as long as it leaves the imports and exports it sees alone.

Reviewer: TypeScript itself is also not a host, because it does not provide any behavior related to modules beyond trying to model other hosts

### Module resolution is host-defined

While the ECMAScript specification defines how to parse and interpret `import` and `export` statements, it leaves module resolution up to the host. If you’re creating a hot new JavaScript runtime, you’re free to create a module resolution scheme like:

```ts
import monkey from "🐒"; // Looks for './eats/bananas.js'
import cow from "🐄";    // Looks for './eats/grass.js'
import lion from "🦁";   // Looks for './eats/you.js'
```

and still claim to implement “standards-compliant ESM.” Needless to say, TypeScript would have no idea what types to assign to `monkey`, `cow`, and `lion` without built-in knowledge of this runtime’s module resolution algorithm. Just as `module` informs the compiler about the host’s expected module format, `moduleResolution`, along with a few customization options, specify the algorithm the host uses to resolve module specifiers to files.

> > #### Why doesn’t TypeScript delegate module resolution to third-party plugins?

Reviewer: I feel like this is probably removable for now

  style main.d.ts stroke-dasharray: 15 4
```

Because of this relationship, the compiler _assumes_ that wherever it sees a declaration file, there is a corresponding JavaScript file that is perfectly described by the type information in the declaration file. For performance reasons, in every module resolution mode, the compiler always looks for TypeScript and declaration files first, and if it finds one, it doesn’t continue looking for the corresponding JavaScript file—if it finds a TypeScript input file, it knows a JavaScript file _will_ exist after compilation, and if it finds a declaration file, it knows a compilation (perhaps someone else’s) already happened and created a JavaScript file at the same time as the declaration file.

“But wait! Plenty of declaration files are written by hand, _not_ generated by `tsc`. Ever heard of DefinitelyTyped?” you might object. And it’s true—hand-writing declaration files, or even moving/copying/renaming them to represent outputs of an external build tool, is a dangerous, error-prone venture. Luckily, there are only three rules that have to be followed to ensure that hand-placed declaration files don’t mislead the module resolution process:

> 1. **Every declaration file represents exactly one JavaScript file.** If you try to use a single declaration file to represent two implementation files—an ES module and a CJS module with similar contents, for example—things can go off the rails. On the other hand, writing a declaration file that represents _zero_ JavaScript files is also dangerous, because it allows an import to be written to a JS file that doesn’t exist. Typically, such an import statement will be erased as long as it only imports types, but can be preserved under `verbatimModuleSyntax`, leading to a runtime error.

Reviewer: How should this be caveated as regards to export maps?

Author: Not sure what you mean?

   {
    "name": "greetings",
    "main": "greetings.js", // Assume these files exist.
    "types": "index.d.ts"   // The declaration file name is different!
   }
   ```
   ```ts
   // @moduleResolution: node16
   // @Filename: /main.ts
   import { sayHello } from "greetings";         // ✅ TS: index.d.ts, Node.js: greetings.js.
   import { sayGoodbye } from "greetings/index"; // ❌ TS: index.d.ts, Node.js: 💥
   ```
   It might have seemed fine for the declaration file to be named `index.d.ts` while the JavaScript file was named `greetings.js` since these filenames are hidden behind `"types"` and `"main"` for root package-name imports. But according to Node.js’s module resolution rules, we’re also allowed to access individual files in this package by name, where `"types"` and `"main"` don’t apply. There are ways to put declaration files in dedicated directories in published packages, but it requires extra care (and configuration) to ensure that the correct mapping from declaration file location to JavaScript file location is always used. It’s highly recommended to colocate declaration files with their implementation files until there’s a good reason not to.

> ### A tour of common module resolution features

Author: Much or all of this will get deleted or replaced with brief mentions and links to the reference pages

If no types are found in a package, TypeScript will make an additional lookup for the package in `node_modules/@types` directories, where community-contributed type definitions from [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) can be installed. The resolution process for a package specifier might go something like this:

```ts
// @Filename: /project/src/main.ts
import { add } from "lodash";
```

- `/project/src/node_modules` doesn’t exist
- `/project/node_modules/lodash/package.json` exists
- No `"typesVersions"`, `"typings"`, or `"types"` field
- The `"main"` field in `package.json` is `"lodash.js"`, but we want to find a declaration file
- `/project/node_modules/lodash/lodash.ts` doesn’t exist
- `/project/node_modules/lodash/lodash.tsx` doesn’t exist
- `/project/node_modules/lodash/lodash.d.ts` doesn’t exist
- `/project/node_modules/lodash/lodash.js.ts` doesn’t exist
- `/project/node_modules/lodash/lodash.js.tsx` doesn’t exist
- `/project/node_modules/lodash/lodash.js.d.ts` doesn’t exist

> - `/project/node_modules/lodash/lodash.js` isn’t a directory

Reviewer: Why?

Author: This section is getting deleted

// - All good 👍
```

However, when Node.js added support for ES modules, they chose to use a stricter module resolution algorithm for imports than they had done for `require` calls. In Node.js, `import` declarations and dynamic `import("...")` calls do not support extension searching or directory `index` lookups. This restriction is respected in TypeScript’s `node16` and `nodenext` resolution modes. However, since the restriction only applies to `import` statements, not `require` calls, TypeScript enforces the restriction only for imports that get _emitted_ as imports. In `--module node16` and `nodenext`, it’s still possible to write import declarations that are transformed to `require` calls, and in these declarations, the looser `require`-based module resolution algorithm is applied:

```ts
// @Filename: main.cts
import { add } from "./math"; // ✅ Ok - emits as `require("./math")` due to .cts extension

// @Filename: main.mts
import { add } from "./math"; // ❌ Needs `.js` extension - emits as-is due to .mts extension
```

### Module resolution for bundlers, TypeScript runtimes, and Node.js loaders

> So far, we’ve really emphasized the distinction between _input files_ and _output files_. Recal that when specifying a file extension on a relative module specifier, TypeScript typically [makes you use the _output_ file extension](#typescript-imitates-the-hosts-module-resolution-but-with-types):

Reviewer: ```suggestion
So far, we’ve really emphasized the distinction between _input files_ and _output files_. Recall that when specifying a file extension on a relative module specifier, TypeScript typically [makes you use the _output_ file extension](#typescript-imitates-the-hosts-module-resolution-but-with-types):
```
